File: backend/app/source_classifier.py
from functools import lru_cache

_UNKNOWN: dict = {"canal": "Inconnu", "sous_canal": "Inconnu"}

_unknown_log: set[str] = set()


@lru_cache(maxsize=2048)
def classify(source: str) -> dict:
    s = str(source or "").lower().strip()
    if not s or s in ("nan", "none", "xxx", "...", "axel", "bannarti", "source-3"):
        return _UNKNOWN

    # --- Paid ---
    if s.startswith("ads_pub_") or s == "ads_retargeting":
        return {"canal": "Paid", "sous_canal": "Meta"}
    if s.startswith("ads2_google") or "googleads" in s or "google-ads" in s or "google ads" in s:
        return {"canal": "Paid", "sous_canal": "Google"}
    if "tiktok" in s:
        return {"canal": "Paid", "sous_canal": "TikTok"}
    if s.startswith("meta") or s.startswith("fb-"):
        return {"canal": "Paid", "sous_canal": "Meta"}

    # --- YouTube ---
    if (
        s.startswith("ytb")
        or s.startswith("yt-")
        or s in ("chaineytb", "post-youtube", "linktree", "linktre")
    ):
        return {"canal": "Organique", "sous_canal": "YouTube"}

    # --- SEO / Blog ---
    if any(
        x in s
        for x in [
            "seo",
            "article",
            "comparatif",
            "banniere_art",
            "modele_art",
            "liste_etf",
            "comment-investir",
            "simulateur",
            "investir1000e",
            "commencez-ici",
            "investissement",
        ]
    ):
        return {"canal": "Organique", "sous_canal": "SEO"}

    # --- Podcast ---
    if any(x in s for x in ["podcast", "ausha", "detente-financiere"]):
        return {"canal": "Organique", "sous_canal": "Podcast"}

    # --- Newsletter ---
    if any(x in s for x in ["substack", "sequencelbd", "mail"]):
        return {"canal": "Organique", "sous_canal": "Newsletter"}

    # --- Webinaire ---
    if s.startswith("webi-"):
        return {"canal": "Organique", "sous_canal": "Webinaire"}

    # --- Social ---
    if any(x in s for x in ["linkedin", "bio-insta", "bio-facebook"]):
        return {"canal": "Organique", "sous_canal": "Social"}

    # --- Affiliation ---
    if s.startswith("aff-"):
        return {"canal": "Organique", "sous_canal": "Affiliation"}

    # --- Owned (pages site) ---
    if any(
        x in s
        for x in [
            "popup",
            "livre",
            "accueilsite",
            "tlmpreb",
            "page-outils",
            "page-parrainage",
            "footer",
            "page-contact",
            "page-a-propos",
            "page-recherche",
            "chatbot",
            "bio-facebook",
            "legend",
        ]
    ):
        return {"canal": "Organique", "sous_canal": "Owned"}

    # --- Direct ---
    if any(x in s for x in ["mon-compte", "pdv", "r2", "direct", "menu", "site"]):
        return {"canal": "Direct", "sous_canal": "Direct"}

    _unknown_log.add(source)
    return _UNKNOWN


def get_unknown_sources() -> list[str]:
    return sorted(_unknown_log)

File: workers/src/lib/source_classifier.ts
/**
 * Port TypeScript de `backend/app/source_classifier.py`.
 *
 * Logique strictement équivalente à la version Python. Les comparaisons
 * `startsWith` et `includes` reproduisent le `s.startswith(...)` et `x in s`
 * du code original.
 */

export interface Classification {
  canal: 'Paid' | 'Organique' | 'Direct' | 'Inconnu'
  sous_canal: string
}

const UNKNOWN: Classification = { canal: 'Inconnu', sous_canal: 'Inconnu' }

const UNKNOWN_SENTINELS = new Set([
  '', 'nan', 'none', 'xxx', '...', 'axel', 'bannarti', 'source-3',
])

const SEO_PATTERNS = [
  'seo', 'article', 'comparatif', 'banniere_art', 'modele_art', 'liste_etf',
  'comment-investir', 'simulateur', 'investir1000e', 'commencez-ici',
  'investissement',
]
const PODCAST_PATTERNS = ['podcast', 'ausha', 'detente-financiere']
const NEWSLETTER_PATTERNS = ['substack', 'sequencelbd', 'mail']
const SOCIAL_PATTERNS = ['linkedin', 'bio-insta', 'bio-facebook']
const OWNED_PATTERNS = [
  'popup', 'livre', 'accueilsite', 'tlmpreb', 'page-outils', 'page-parrainage',
  'footer', 'page-contact', 'page-a-propos', 'page-recherche', 'chatbot',
  'bio-facebook', 'legend',
]
const DIRECT_PATTERNS = ['mon-compte', 'pdv', 'r2', 'direct', 'menu', 'site']

const cache = new Map<string, Classification>()
const unknownLog = new Set<string>()

export function classify(source: string | null | undefined): Classification {
  const original = source == null ? '' : String(source)
  if (cache.has(original)) return cache.get(original)!

  const s = original.toLowerCase().trim()
  if (UNKNOWN_SENTINELS.has(s)) {
    cache.set(original, UNKNOWN)
    return UNKNOWN
  }

  let result: Classification

  if (s.startsWith('ads_pub_') || s === 'ads_retargeting') {
    result = { canal: 'Paid', sous_canal: 'Meta' }
  } else if (s.startsWith('ads2_google') || s.includes('googleads') || s.includes('google-ads') || s.includes('google ads')) {
    result = { canal: 'Paid', sous_canal: 'Google' }
  } else if (s.includes('tiktok')) {
    result = { canal: 'Paid', sous_canal: 'TikTok' }
  } else if (s.startsWith('meta') || s.startsWith('fb-')) {
    result = { canal: 'Paid', sous_canal: 'Meta' }
  } else if (
    s.startsWith('ytb') ||
    s.startsWith('yt-') ||
    s === 'chaineytb' || s === 'post-youtube' || s === 'linktree' || s === 'linktre'
  ) {
    result = { canal: 'Organique', sous_canal: 'YouTube' }
  } else if (SEO_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Organique', sous_canal: 'SEO' }
  } else if (PODCAST_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Organique', sous_canal: 'Podcast' }
  } else if (NEWSLETTER_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Organique', sous_canal: 'Newsletter' }
  } else if (s.startsWith('webi-')) {
    result = { canal: 'Organique', sous_canal: 'Webinaire' }
  } else if (SOCIAL_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Organique', sous_canal: 'Social' }
  } else if (s.startsWith('aff-')) {
    result = { canal: 'Organique', sous_canal: 'Affiliation' }
  } else if (OWNED_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Organique', sous_canal: 'Owned' }
  } else if (DIRECT_PATTERNS.some((x) => s.includes(x))) {
    result = { canal: 'Direct', sous_canal: 'Direct' }
  } else {
    unknownLog.add(original)
    result = UNKNOWN
  }

  cache.set(original, result)
  return result
}

export function getUnknownSources(): string[] {
  return [...unknownLog].sort()
}
